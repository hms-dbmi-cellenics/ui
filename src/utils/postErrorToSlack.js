import Auth from '@aws-amplify/auth';
import StackTrace from 'stacktrace-js';
import { getLoggerBotToken } from 'utils/slack';

const extractExperimentId = (url) => {
  const match = url.match(/experiments\/([^/]+)/i);
  return match ? match[1] : null;
};

// Truncates data arrays so that it doesn't produce large logs
const NUM_DATA_TO_SHOW = 20;
const truncateCollection = (arr) => {
  if (arr.length < NUM_DATA_TO_SHOW) return arr;

  const truncatedArr = arr.slice(0, NUM_DATA_TO_SHOW);
  truncatedArr.push('...');
  return truncatedArr;
};

const trimOutput = (key, item) => {
  if (Array.isArray(item)) return truncateCollection(item);
  if (item instanceof Set) return new Set(truncateCollection(Array.from(item)));

  if (item !== null && typeof item === 'object') {
    const newItem = Object.keys(item).reduce((newObject, childKey) => {
      // eslint-disable-next-line no-param-reassign
      newObject[childKey] = trimOutput(childKey, item[childKey]);
      return newObject;
    }, {});

    return newItem;
  }

  return item;
};

const buildErrorMessage = async (errorObject, reduxState, context) => {
  const {
    user, timestamp, experimentId, url,
  } = context;

  const stack = await StackTrace.fromError(errorObject);
  const stringifiedStack = stack.map((sf) => sf.toString()).join('\n');

  let message = `
    ===== ERROR STACK =====
    Message:
    ${errorObject.message}
    At:
    ${stringifiedStack}

    === DETAILS ===
    User: ${user.attributes.name} <${user.attributes.email}> ${user.username}
    ExperimentID: ${experimentId}
    URL: ${url}
    Timestamp: ${timestamp}

    `;

  if (reduxState) {
    message += `===== REDUX STATE =====
    ${JSON.stringify(reduxState, trimOutput, 2)}`;
  }
  return message;
};

const postError = async (errorLog, context) => {
  const {
    user, timestamp, experimentId, url, environment,
  } = context;

  const message = `
  \u26A0  [${environment.toUpperCase()}] Uncaught UI Error - ExpID ${experimentId} - ${timestamp}
  URL: ${url}

  User: ${user.attributes.name} <${user.attributes.email}> ${user.username}
  Experiment ID: ${experimentId}`;

  const formData = new FormData();

  const formFields = {
    token: getLoggerBotToken(),
    title: `UI Error at ${new Date().toISOString()}`,
    initial_comment: message,
    channels: 'error-logs',
    filename: `UI-error-${experimentId}-${timestamp}.log`,
    content: errorLog,
  };

  Object.entries(formFields).forEach(([fieldName, value]) => {
    formData.append(fieldName, value);
  });

  try {
    const res = await fetch('https://slack.com/api/files.upload', {
      method: 'POST',
      body: formData,
    });

    if (!res.ok) {
      throw new Error(
        `Failed sending error message to Slack: ${res.status}`, res.statusText,
      );
    }
  } catch (err) {
    console.error(err);
  }
};

const postErrorToSlack = async (errorObject, reduxState) => {
  const user = await Auth.currentAuthenticatedUser();

  const timestamp = new Date().toISOString();
  const url = window.location.href;
  const experimentId = extractExperimentId(url);
  const { networkResources: { environment } } = reduxState;

  const context = {
    user,
    timestamp,
    experimentId,
    url,
    environment,
  };

  const errorLog = await buildErrorMessage(errorObject, reduxState, context);
  await postError(errorLog, context);
};

export default postErrorToSlack;
