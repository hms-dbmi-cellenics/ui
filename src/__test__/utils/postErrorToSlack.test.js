import postErrorToSlack from 'utils/postErrorToSlack';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

enableFetchMocks();

const mockError = { message: 'mockMessage', stack: 'Mock error stack' };
const mockReduxDump = {
  cellInfo: {
    focus: {
      store: 'cellSets',
      key: 'louvain',
    },
    groupedTrack: 'louvain',
    selectedTracks: [
      'louvain',
    ],
  },
  cellSets: {
    properies: {
      'louvain-0': {
        cellIds: new Set(Array(25).fill(0)),
      },
    },
  },
  cellMeta: {
    mitochondrialContent: {
      loading: true,
      error: false,
      data: new Array(25).fill(0),
    },
  },
  networkResources: {
    environment: 'test',
  },
};

jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

delete window.location;
window.location = {
  href: 'http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration',
};

jest.mock('@aws-amplify/auth', () => ({
  currentAuthenticatedUser: () => Promise.resolve({
    attributes: {
      name: 'John Doe',
      email: 'fake@email.com',
    },
    username: '5152fake-eb52-474c-user-mocke8c8user',
  }),
}));

jest.mock('stacktrace-js', () => ({
  fromError: () => Promise.resolve(['line1', 'line2']),
}));

describe('PostErrorToSlack', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fetchMock.resetMocks();
  });

  it('Posts requests correctly', async () => {
    await postErrorToSlack(mockError, mockReduxDump);

    expect(fetchMock).toHaveBeenCalledTimes(1);
    expect(fetchMock.mock.calls[0]).toMatchSnapshot();

    const { body: formData } = fetchMock.mock.calls[0][1];

    // Check content of body
    formData.forEach((value, key) => {
      // Skip if key === token because it contains a token
      if (key === 'token') return;

      expect(value).toMatchSnapshot();
    });
  });

  it('Should not throw an error if POSTing fails', async () => {
    fetchMock.mockIf(/.*/, () => Promise.resolve(new Response('Server error', { status: 500 })));

    await postErrorToSlack(mockError, mockReduxDump);

    expect(fetchMock).toHaveBeenCalledTimes(1);
  });
});
