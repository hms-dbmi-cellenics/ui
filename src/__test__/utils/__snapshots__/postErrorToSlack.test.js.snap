// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostErrorToSlack Posts requests correctly 1`] = `
Array [
  "https://slack.com/api/files.upload",
  Object {
    "body": FormData {},
    "method": "POST",
  },
]
`;

exports[`PostErrorToSlack Posts requests correctly 2`] = `"UI Error at 2020-01-01T00:00:00.000Z"`;

exports[`PostErrorToSlack Posts requests correctly 3`] = `
"
  âš   [TEST] Uncaught UI Error - ExpID testae48e318dab9a1bd0bexperiment - 2020-01-01T00:00:00.000Z
  URL: http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration

  User: John Doe <fake@email.com> 5152fake-eb52-474c-user-mocke8c8user
  Experiment ID: testae48e318dab9a1bd0bexperiment"
`;

exports[`PostErrorToSlack Posts requests correctly 4`] = `"error-logs"`;

exports[`PostErrorToSlack Posts requests correctly 5`] = `"UI-error-testae48e318dab9a1bd0bexperiment-2020-01-01T00:00:00.000Z.log"`;

exports[`PostErrorToSlack Posts requests correctly 6`] = `
"
    ===== ERROR STACK =====
    Message:
    mockMessage
    At:
    line1
line2

    === DETAILS ===
    User: John Doe <fake@email.com> 5152fake-eb52-474c-user-mocke8c8user
    ExperimentID: testae48e318dab9a1bd0bexperiment
    URL: http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration
    Timestamp: 2020-01-01T00:00:00.000Z

    ===== REDUX STATE =====
    {
  \\"cellInfo\\": {
    \\"focus\\": {
      \\"store\\": \\"cellSets\\",
      \\"key\\": \\"louvain\\"
    },
    \\"groupedTrack\\": \\"louvain\\",
    \\"selectedTracks\\": [
      \\"louvain\\"
    ]
  },
  \\"cellSets\\": {
    \\"properies\\": {
      \\"louvain-0\\": {
        \\"cellIds\\": {}
      }
    }
  },
  \\"cellMeta\\": {
    \\"mitochondrialContent\\": {
      \\"loading\\": true,
      \\"error\\": false,
      \\"data\\": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        \\"...\\"
      ]
    }
  },
  \\"networkResources\\": {
    \\"environment\\": \\"test\\"
  }
}"
`;
